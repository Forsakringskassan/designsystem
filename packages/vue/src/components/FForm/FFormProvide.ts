import { Reference } from "@fkui/logic";
import { FormErrorList } from "../../types";

/**
 * @public
 */
export const setRef = Symbol("setRef");

/**
 * @public
 */
export const getRef = Symbol("getRef");

/**
 * @public
 */
export const setIsOpen = Symbol("setIsOpen");

/**
 * Methods provided by FForm.
 *
 * @public
 */
export interface FFormProvider {
    [setRef](id: string, data: FormErrorList): void;
    [getRef](id: string): Reference<FormErrorList>;
}

/**
 * @public
 */
export function createFFormProvideOptions(vm: {
    components: Record<string, Reference<FormErrorList>>;
}): FFormProvider {
    const { components } = vm;
    return {
        [setRef](id: string, data: FormErrorList): void {
            components[id] = new Reference(data);
        },
        [getRef](id: string): Reference<FormErrorList> {
            return components[id];
        },
    };
}
