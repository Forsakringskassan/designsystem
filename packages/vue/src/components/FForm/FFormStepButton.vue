<template>
    <button
        :id="id"
        class="button button--margin-bottom-0 button--large"
        type="button"
        v-bind="$attrs"
        :aria-expanded="isOpen ? 'true' : 'false'"
        data-form-step-button
    >
        <template v-if="isOpen">
            <f-icon name="success" class="button__icon" />
            {{ $t("fkui.form-step-button.close", "OK") }}
        </template>
        <template v-else>
            <f-icon name="pen" class="button__icon" />
            {{
                isAnyFieldTouched
                    ? $t("fkui.form-step-button.edit", "Ã„ndra")
                    : $t("fkui.form-step-button.open", "Fyll i")
            }}
        </template>
        <span class="sr-only">{{ additionalScreenreaderText }}</span>
    </button>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FIcon } from "../FIcon";
import { TranslationMixin } from "../../plugins";

/**
 * @deprecated `<f-form-step-button>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/components/fform.html
 */
export default defineComponent({
    name: "FFormStepButton",
    components: { FIcon },
    mixins: [TranslationMixin],
    inheritAttrs: false,
    model: {
        prop: "sr-only",
    },
    props: {
        /**
         * If the id is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        isOpen: Boolean,
        isAnyFieldTouched: Boolean,
        additionalScreenreaderText: {
            type: String,
            required: false,
            default: "",
        },
    },
});
</script>
