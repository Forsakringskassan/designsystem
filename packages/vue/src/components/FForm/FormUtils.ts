import { Reference, documentOrderComparator } from "@fkui/logic";
import { ComponentValidityEvent, FormErrorList, FormStep } from "../../types";

/**
 * @public
 */
export type ComponentReferences =
    | Record<string, ComponentValidityEvent>
    | Record<string, Reference<FormErrorList>>;

/**
 * @public
 */
export type ComponentValueTypes =
    | ComponentValidityEvent
    | Reference<FormErrorList>;

/**
 * Clean inputs which doesn't exist in the DOM
 *
 * @public
 */
export function cleanUpElements(vm: {
    $el: Element;
    components: ComponentReferences;
}): Promise<void> {
    return new Promise((resolve) => {
        /**
         * To be able to clean up validation errors after elements that have been v-if:ed away (removed from DOM)
         * the clean up needs to be runned after DOM has been updated, thus setTimeout. $nextTick can't be
         * used due to only virtual DOM in Vue has been updated and not actual browse DOM.
         */
        window.setTimeout(() => {
            Object.keys(vm.components).forEach((id) => {
                const domElement = vm.$el.querySelector(`#${id}`);
                if (!domElement) {
                    /* eslint-disable-next-line @typescript-eslint/no-dynamic-delete -- technical debt, this is potentially dangerous, should probably use a safer structure such as Map */
                    delete vm.components[id];
                }
            });
            resolve();
        }, 0);
    });
}

/**
 * Sorts errorlist based on the order of the elements in the DOM
 *
 * @public
 */
export function sortComponentsWithErrorsOnDOMOrder(
    componentList: ComponentReferences,
): ComponentValueTypes[] {
    const errorList = Object.values(componentList).filter((component) => {
        const validity =
            component instanceof Reference
                ? component.ref.isValid
                : component.validityMode;
        return typeof validity === "boolean"
            ? validity === false
            : validity === "ERROR";
    }) as ComponentValueTypes[];

    errorList.sort((a: ComponentValueTypes, b: ComponentValueTypes) => {
        const elementToCompareA =
            a instanceof Reference
                ? document.querySelector(`#${a.ref.id}`)
                : a.target;
        const elementToCompareB =
            b instanceof Reference
                ? document.querySelector(`#${b.ref.id}`)
                : b.target;
        return documentOrderComparator(elementToCompareA, elementToCompareB);
    });
    return errorList;
}

/**
 * Checks if reference is FormStep
 *
 * @public
 */
export function isFormStepReference(
    reference: Reference<FormErrorList | FormStep>,
): reference is Reference<FormStep> {
    return "isOpen" in reference.ref;
}
