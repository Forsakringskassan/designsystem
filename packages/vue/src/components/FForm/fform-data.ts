import { Reference } from "@fkui/logic";
import { FormErrorList } from "../../types";

/**
 * @public
 */
export interface FFormData {
    errorMessageSlotClass: string[];
    components: Record<string, Reference<FormErrorList>>;
}
