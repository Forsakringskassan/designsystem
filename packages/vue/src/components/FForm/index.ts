export { default as FForm } from "./FForm.vue";
export { default as FFormStep } from "./FFormStep.vue";
export {
    type FFormProvider,
    setRef,
    getRef,
    setIsOpen,
    createFFormProvideOptions,
} from "./FFormProvide";
export {
    type ComponentReferences,
    type ComponentValueTypes,
    cleanUpElements,
    isFormStepReference,
    sortComponentsWithErrorsOnDOMOrder,
} from "./FormUtils";
export { type FFormData } from "./fform-data";
