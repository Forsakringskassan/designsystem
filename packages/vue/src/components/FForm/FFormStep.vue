<template>
    <f-validation-group
        :id="id"
        v-bind="attrs"
        :key="validationGroupKey"
        v-model="groupValidity"
        class="form-step"
        :stop-propagation="true"
    >
        <div class="form-step__header" tabindex="-1" role="group">
            <!-- @slot
Slot for header content, e.g. title.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ slotClass }"`.

The following properties are available:

* `slotClass: string[];` CSS classes to use for the heading element. Use with `:class="slotClass"`.
* `+` all properties that are available in the default slot.
      -->
            <slot name="header" v-bind="{ slotClass: headerSlotClass, ...self.ref, toggleIsOpen }"></slot>
            <f-icon v-if="self.ref.isValid" class="form-step__check" name="success"></f-icon>
        </div>
        <div v-if="displayErrorList" :id="stepErrorId" class="form-step__error" tabindex="-1" role="group">
            <nav>
                <f-error-list :bullets="true" :items="componentsWithError" :before-navigate="beforeNavigateToError">
                    <template #title>
                        <!--
                             @slot Slot for displaying error description.

                             After this slot a list of invalid elements is listed.
                             When an item is clicked it will scroll to and focus that invalid element.
                        -->
                        <slot name="error-message">
                            {{ defaultErrorMessage }}
                        </slot>
                    </template>
                </f-error-list>
            </nav>
        </div>

        <div @component-validity="onComponentValidity">
            <!-- @slot
Slot for displaying content, e.g. descriptions and input elements.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ isOpen, isSubmitted }"`.

The following properties are available:

* `isOpen: boolean;`
* `isSubmitted: boolean;`
* `isAnyFieldTouched: boolean;`
* `isValid: boolean;`
* `id: string;`
* `title: string;`
* `toggleIsOpen: () => void;`
    -->
            <slot name="default" v-bind="{ ...self.ref, isSubmitted, toggleIsOpen }"></slot>
            <span v-if="hasArrow" class="form-step__arrow"></span>
        </div>
    </f-validation-group>
</template>

<script lang="ts">
import { defineComponent, inject } from "vue";
import { ElementIdService, ValidationService, focus, Reference } from "@fkui/logic";
import { FormStep, ErrorItem, GroupValidityEvent, ComponentValidityEvent } from "../../types";
import { FIcon } from "../FIcon";
import { FErrorList } from "../FErrorList";
import { FValidationGroup } from "../FValidationGroup";
import { TranslationMixin } from "../../plugins";
import { setRef, getRef, setIsOpen } from "./FFormProvide";

/**
 * @deprecated `<f-form-step>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/components/fform.html
 */
export default defineComponent({
    name: "FFormStep",
    components: { FIcon, FValidationGroup, FErrorList },
    mixins: [TranslationMixin],
    inheritAttrs: false,
    props: {
        /**
         * The id for the root div id attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: true,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * If the form step should have a bottom arrow.
         * It is recommended to remove the arrow for the last step with `:has-arrow="false"`.
         */
        hasArrow: {
            type: Boolean,
            default: true,
        },
        /**
         * If the form step is the last step.
         * If set to true the form step will have a bigger bottom margin.
         */
        isLastStep: {
            type: Boolean,
            default: false,
        },
        /**
         * If error links should be disabled, `:disable-error-links="true"`..
         * If `true`, errors will be displayed as text instead.
         */
        disableErrorLinks: {
            type: Boolean,
            default: false,
        },
    },
    setup() {
        return {
            setRef: inject(setRef) as (id: string, data: FormStep | FormStep) => void,
            getRef: inject(getRef) as (id: string) => Reference<FormStep | FormStep>,
            setIsOpen: inject(setIsOpen) as (id: string, isOpen: boolean) => void,
        };
    },
    data() {
        return {
            self: new Reference<FormStep>({
                isOpen: false,
                numberOfTimesSubmitted: 0,
                isAnyFieldTouched: false,
                isValid: true,
                id: this.id,
                focusElementId: `${this.id}-error`,
                title: "",
            }),
            hasBeenValid: false,
            isOpenedClosed: false,
            headerSlotClass: ["form-step__title"],
            stepErrorId: `${this.id}-error`,
            groupValidity: {
                isValid: false,
                componentsWithError: [],
                componentCount: 0,
            } as GroupValidityEvent,
        };
    },
    computed: {
        attrs(): Record<string, string | Record<string, unknown>> {
            return {
                ...this.$attrs,
                class: {
                    "form-step--complete": this.self.ref.isValid,
                    "form-step--last-step": this.isLastStep,
                },
            };
        },
        defaultErrorMessage(): string {
            return this.$t("fkui.form-step.errorlist.title", "Oj, du har glömt att fylla i något. Gå till:");
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- technical debt, not sure what is going on here
        injected(): any {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- technical debt, not sure what is going on here
            return this as any;
        },
        isSubmitted(): boolean {
            return this.self.ref.numberOfTimesSubmitted > 0;
        },
        componentsWithError(): ErrorItem[] {
            return this.groupValidity.componentsWithError.map((c: ComponentValidityEvent) => ({
                id: this.disableErrorLinks ? undefined : c.elementId,
                focusElementId: c.focusElementId,
                title: c.errorMessage,
            }));
        },
        hasError(): boolean {
            return this.groupValidity.componentsWithError.length > 0;
        },
        displayErrorList(): boolean {
            return (this.isOpenedClosed || this.hasBeenValid || this.isSubmitted) && this.hasError;
        },
        validationGroupKey(): string {
            return `${this.id}-group`;
        },
    },
    watch: {
        "self.ref.numberOfTimesSubmitted": {
            handler(): void {
                ValidationService.setSubmitted(this.$el);
                ValidationService.validateAllElements(this.id);
            },
        },
        groupValidity: {
            immediate: true,
            handler(): void {
                this.self.ref.isValid = this.groupValidity.isValid;

                if (this.hasBeenValid === false) {
                    this.hasBeenValid = this.groupValidity.isValid;
                }
            },
        },
    },
    async mounted(): Promise<void> {
        await this.$nextTick();
        this.loadSelf();
        this.injected.setRef(this.id, this.self.ref);
        this.loadSelf();
        this.updateSelfRefTitle();
    },
    beforeUpdate() {
        this.updateSelfRefTitle();
    },
    methods: {
        updateSelfRefTitle(): void {
            const [firstSlotClass] = this.headerSlotClass;
            const headerElement = this.$el.querySelector(`.${firstSlotClass}`);
            if (headerElement && headerElement.textContent) {
                this.self.ref.title = headerElement.textContent;
            }
        },
        loadSelf(): void {
            const ref = this.injected.getRef(this.id);
            Object.assign(this.self, ref);
        },
        async focusDomElement(): Promise<void> {
            await this.$nextTick();

            if (this.hasError && this.self.ref.isAnyFieldTouched) {
                const formStepError = this.$el.querySelector(".form-step__error");
                focus(formStepError);
            } else {
                const formStepHeader = this.$el.querySelector(".form-step__header");
                focus(formStepHeader);
            }
        },
        async beforeNavigateToError(): Promise<void> {
            this.injected.setIsOpen(this.id, true);
            await this.$nextTick();
        },
        /**
         * Toggle if open or closed.
         * If another form step is open then it will be closed.
         * @public
         */
        toggleIsOpen(): void {
            this.injected.setIsOpen(this.id, !this.self.ref.isOpen);
            // ? If any field is touched and form step was opened and closed.
            if (this.self.ref.isAnyFieldTouched && this.self.ref.isOpen === false) {
                if (this.isOpenedClosed === false) {
                    this.isOpenedClosed = true;
                }

                ValidationService.setTouched(this.$el);
                ValidationService.validateAllElements(this.id);
            }

            this.focusDomElement();
        },
        onComponentValidity(event: CustomEvent<ComponentValidityEvent>): void {
            const hasReceivedValidOrErrorEvent =
                event.detail.validityMode === "VALID" || event.detail.validityMode === "ERROR";

            if (hasReceivedValidOrErrorEvent && this.self.ref.isOpen) {
                this.self.ref.isAnyFieldTouched = true;
            }
        },
    },
});
</script>
