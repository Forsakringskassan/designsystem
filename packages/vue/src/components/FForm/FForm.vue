<template>
    <!-- [html-validate-disable-next wcag/h32: FForm does not have a default button and leaves the responsiblity to the application using FForm.] -->
    <form :id="id" class="form" novalidate v-bind="attrs" @component-validity="onComponentValidity">
        <f-message-box v-if="displayError && isSubmitted && hasError" tabindex="-1" type="error">
            <!--
@slot Slot for displaying error description.

After this slot a list of invalid elements is listed.
When an item is clicked it will scroll to that invalid element.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ slotClass }"`.

The following properties are available:

* `slotClass: string[];` CSS classes to use for the error message heading element. Use with `:class="slotClass"`.
      -->
            <slot name="error-message" v-bind="{ slotClass: errorMessageSlotClass }"></slot>
            <nav>
                <f-error-list :items="componentsWithErrors"></f-error-list>
            </nav>
        </f-message-box>
        <!-- @slot Slot for displaying content, e.g. input elements and form steps. -->
        <slot name="default"></slot>
    </form>
</template>

<script lang="ts">
import { ElementIdService, focus, Reference, ValidationService } from "@fkui/logic";
import { defineComponent, type PropType } from "vue";
import { ComponentValidityEvent, FormErrorList, FormStep } from "../../types";
import { FMessageBox } from "../FMessageBox";
import { FErrorList } from "../FErrorList";
import { type FFormProvider, createFFormProvideOptions } from "./FFormProvide";
import { cleanUpElements, isFormStepReference, sortComponentsWithErrorsOnDOMOrder } from "./FormUtils";

interface FFormData {
    errorMessageSlotClass: string[];
    components: Record<string, Reference<FormErrorList | FormStep>>;
}

/**
 * @deprecated `<f-form>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/components/fform.html
 */
export default defineComponent({
    name: "FForm",
    components: { FMessageBox, FErrorList },
    provide(): FFormProvider {
        return createFFormProvideOptions(this);
    },
    inheritAttrs: false,
    props: {
        /**
         * The id for the form id attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * If the error-message slot and the invalid elements list
         * should be displayed.
         */
        displayError: {
            type: Boolean,
            default: true,
        },
        /**
         * If invalid fields at submit this tells how the scroll to the error
         * field should occur. The default, "center", centres the error information
         * vertically. Allowed value is also "top" which scrolls such that the error
         * informations top is at the top of the viewport.
         */
        errorScroll: {
            type: String as PropType<"center" | "top">,
            default: "center",
            validator(value: string) {
                return ["center", "top"].includes(value);
            },
        },
    },
    data(): FFormData {
        return {
            errorMessageSlotClass: ["message-box__heading"],
            components: {},
        };
    },
    computed: {
        attrs(): Record<string, unknown> {
            type SubmitHandler = (e: SubmitEvent) => void;
            let onSubmit: SubmitHandler | SubmitHandler[];
            // ? Merge submit listeners
            if (this.$attrs.onSubmit) {
                const originalSubmit = this.$attrs.onSubmit as SubmitHandler | SubmitHandler[];

                onSubmit = async (event: SubmitEvent) => {
                    const isValid = await this.onSubmit(event);
                    if (isValid) {
                        if (Array.isArray(originalSubmit)) {
                            originalSubmit.forEach((submitFunction) => submitFunction(event));
                        } else {
                            originalSubmit(event);
                        }
                    }
                };
            } else {
                onSubmit = this.onSubmit;
            }
            const attrs: Record<string, unknown> = { ...this.$attrs, id: this.id, onSubmit };
            return attrs;
        },
        numberOfTimesSubmitted(): number {
            const components: Array<Reference<FormErrorList | FormStep>> = Object.values(this.components);
            return Math.max(0, ...components.map((it) => it.ref.numberOfTimesSubmitted));
        },
        isSubmitted(): boolean {
            return this.numberOfTimesSubmitted > 0;
        },
        hasError(): boolean {
            if (Object.values(this.components).length === 0) {
                return false;
            } else {
                const components: Array<Reference<FormErrorList | FormStep>> = Object.values(this.components);
                return components.some((component) => component.ref.isValid === false);
            }
        },
        componentsWithErrors(): Array<FormErrorList | FormStep> {
            const sortedComponents = sortComponentsWithErrorsOnDOMOrder(this.components) as Array<
                Reference<FormErrorList | FormStep>
            >;

            return sortedComponents.map((c) => c.ref);
        },
        needsErrorScroll(): boolean {
            return this.errorScroll !== "center";
        },
    },
    methods: {
        async focusDomElement(): Promise<void> {
            await this.$nextTick();
            const messageBoxError = this.$el.querySelector(".message-box--error");
            if (messageBoxError) {
                focus(messageBoxError, { scrollToTop: this.needsErrorScroll });
            }
        },
        async focusFallbackElement(): Promise<void> {
            await this.$nextTick();
            const element = this.$el.querySelector(".form-step__header");
            const invalidElement = this.$el.querySelector(":invalid");
            if (element) {
                focus(element, { scrollToTop: this.needsErrorScroll });
            } else if (invalidElement) {
                focus(invalidElement, { force: true, scrollToTop: this.needsErrorScroll });
            }
        },
        async onSubmit(event: Event): Promise<boolean> {
            event.preventDefault();

            const components: Array<Reference<FormErrorList | FormStep>> = Object.values(this.components);
            components.forEach((component) => {
                component.ref.numberOfTimesSubmitted++;

                if (!isFormStepReference(component)) {
                    ValidationService.setTouched(component.ref.id);
                    ValidationService.setSubmitted(component.ref.id);
                }
            });

            await cleanUpElements(this);
            await ValidationService.validateAllElements(this.id);

            if (this.displayError) {
                this.focusDomElement();
            } else {
                // ! TODO: Focus on first invalid FormStep error list.
                this.focusFallbackElement();
            }

            return this.hasError === false;
        },
        async onComponentValidity(event: CustomEvent<ComponentValidityEvent>): Promise<void> {
            const reference = new Reference<FormErrorList | FormStep>({
                id: event.detail.elementId,
                focusElementId: event.detail.focusElementId,
                title: event.detail.errorMessage,
                isValid: event.detail.isValid,
                numberOfTimesSubmitted: this.numberOfTimesSubmitted,
            });

            this.components[event.detail.elementId] = reference;

            // required for components not within formsteps
            await cleanUpElements(this);
        },
    },
});
</script>
