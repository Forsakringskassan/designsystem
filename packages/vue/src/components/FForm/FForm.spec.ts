import "html-validate/jest";
import "@fkui/test-utils/jest";
import flushPromises from "flush-promises";
import { defineComponent } from "vue";
import { VueWrapper, mount } from "@vue/test-utils";
import { ValidationService, Reference } from "@fkui/logic";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import { FTextField } from "../FTextField";
import { FormStep, FormErrorList } from "../../types";
import { ValidationPlugin } from "../../plugins";
import FForm from "./FForm.vue";

jest.useFakeTimers();

const TestComponent = defineComponent({
    components: { FForm, FTextField },
    template: /* HTML */ `
        <f-form v-bind="$attrs">
            <template #error-message>
                <div>error-message slot content</div>
            </template>
            <template #default>
                <f-text-field id="element1" v-validation.required>
                    Element 1
                </f-text-field>
                <f-text-field id="element2" v-validation.required>
                    Element 2
                </f-text-field>
            </template>
        </f-form>
    `,
});

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
    stubs = {},
} = {}): VueWrapper {
    return mount(TestComponent, {
        attrs: { ...attrs },
        props: { id: "test-id", ...props },
        slots: { ...slots },
        attachTo: createPlaceholderInDocument(),
        global: {
            plugins: [ValidationPlugin],
            stubs: {
                FLabel: true,
                FIcon: true,
                FErrorList: {
                    props: ["items"],
                    template: /* HTML */ `
                        <div>
                            <pre v-for="item of items" v-text="item"></pre>
                        </div>
                    `,
                },
                ...stubs,
            },
        },
    });
}

async function submitForm(wrapper: VueWrapper): Promise<void> {
    await wrapper.get("form").trigger("submit");
    jest.runAllTimers();
    await flushPromises();
}

async function fillForm(wrapper: VueWrapper): Promise<void> {
    const inputs = wrapper.findAll("input");
    for (const input of inputs) {
        await input.trigger("focus");
    }
    await wrapper.vm.$nextTick();
    for (const input of inputs) {
        await input.setValue("foobar");
    }
    await wrapper.vm.$nextTick();
    for (const input of inputs) {
        await input.trigger("blur");
    }
    await wrapper.vm.$nextTick();
    await flushPromises();
}

function getErrors(wrapper: VueWrapper): unknown[] {
    return wrapper
        .findAll("div pre")
        .map((it) => JSON.parse(it.element.textContent ?? "{}"));
}

afterEach(() => {
    jest.restoreAllMocks();
});

describe("error list", () => {
    it.todo("should not display error list when there are no errors");
    it.todo("should not display error list when disabled");
    it.todo("should display error list when enabled");
});

describe("snapshots", () => {
    it("should match snapshot when pristine", async () => {
        expect.assertions(1);
        const wrapper = createWrapper();
        await flushPromises();
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with errors", async () => {
        expect.assertions(1);
        const wrapper = createWrapper({ stubs: { FMessageBox: true } });
        await flushPromises();
        await submitForm(wrapper);
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("submit", () => {
    it("should increment numberOfTimesSubmitted when submitting", async () => {
        expect.assertions(1);

        /* given: form with input fields */
        const wrapper = createWrapper();
        await flushPromises();

        /* when: form is submitted */
        await submitForm(wrapper);

        const fForm = wrapper.getComponent(FForm);
        const components = fForm.vm.$data.components;
        const numberOfTimesSubmitted = Object.entries(components).map(
            ([name, component]) => {
                const { ref } = component as Reference<
                    FormErrorList | FormStep
                >;
                return [name, ref.numberOfTimesSubmitted];
            },
        );

        expect(numberOfTimesSubmitted).toEqual([
            ["element1", 1],
            ["element2", 1],
        ]);
    });

    it("should emit submit event when valid form is submitted", async () => {
        expect.assertions(1);

        /* given: form with submit listener */
        const submit = jest.fn();
        const wrapper = createWrapper({
            attrs: { onSubmit: submit },
        });
        await flushPromises();

        /* when: valid form is submitted */
        await fillForm(wrapper);
        await submitForm(wrapper);

        expect(submit).toHaveBeenCalled();
    });

    it("should not emit submit event when invalid form is submitted", async () => {
        expect.assertions(1);

        /* given: form with submit listener */
        const submit = jest.fn();
        const wrapper = createWrapper({
            attrs: { onSubmit: submit },
        });
        await flushPromises();

        /* when: invalid form is submitted */
        await submitForm(wrapper);

        expect(submit).not.toHaveBeenCalled();
    });

    it("should set all components to submitted before validation", async () => {
        const setSubmitted = jest.spyOn(ValidationService, "setSubmitted");
        const validateAllElements = jest.spyOn(
            ValidationService,
            "validateAllElements",
        );

        const wrapper = createWrapper();
        await flushPromises();

        await submitForm(wrapper);

        const setSubmittedCallOrder = setSubmitted.mock.invocationCallOrder[0];
        const validateAllElementsCallOrder =
            validateAllElements.mock.invocationCallOrder[0];

        expect(setSubmittedCallOrder).toBeLessThan(
            validateAllElementsCallOrder,
        );
        expect(setSubmitted).toHaveBeenCalledWith("element1");
        expect(setSubmitted).toHaveBeenCalledWith("element2");
    });

    it("should show error when component validity events from new element are dispatched after first submit", async () => {
        expect.assertions(2);

        const wrapper = createWrapper({});
        await flushPromises();

        const element1 = wrapper.get("#element1");
        const element2 = wrapper.get("#element2");

        element1.setValue("");
        element2.setValue("foo");
        await submitForm(wrapper);

        expect(getErrors(wrapper)).toEqual([
            {
                focusElementId: "element1",
                id: "element1",
                isValid: false,
                numberOfTimesSubmitted: 1,
                title: "Element 1",
            },
        ]);

        element2.setValue("");
        await submitForm(wrapper);

        expect(getErrors(wrapper)).toEqual([
            {
                focusElementId: "element1",
                id: "element1",
                isValid: false,
                numberOfTimesSubmitted: 2,
                title: "Element 1",
            },
            {
                focusElementId: "element2",
                id: "element2",
                isValid: false,
                numberOfTimesSubmitted: 2,
                title: "Element 2",
            },
        ]);
    });

    describe("focus", () => {
        it("should focus .message-box--error element if form is invalid", async () => {
            const wrapper = createWrapper();
            await flushPromises();

            await submitForm(wrapper);

            const expectedFocus = wrapper.get(".message-box--error");
            expect(expectedFocus.element).toHaveFocus();
        });

        it("should focus first invalid element if form is invalid with no .form-step__header element and prop displayError = false", async () => {
            const wrapper = createWrapper({
                props: {
                    displayError: false,
                },
            });
            await flushPromises();

            await submitForm(wrapper);

            const expectedFocus = wrapper.get(":invalid");
            expect(expectedFocus.element).toHaveFocus();
        });
    });
});

describe("html-validate", () => {
    it.each`
        html
        ${'<f-form><button type="submit">Submit</button></f-form>'}
        ${'<f-form display-error><button type="submit">Submit</button></f-form>'}
        ${'<f-form display-error="true"><button type="submit">Submit</button></f-form>'}
        ${'<f-form display-error="false"><button type="submit">Submit</button></f-form>'}
        ${'<f-form><template #error-message>Error</template><button type="submit">Submit</button></f-form>'}
    `("$html should be valid", ({ html }) => {
        expect.assertions(1);

        expect(html).toHTMLValidate();
    });

    it.each`
        html
        ${"<f-form></f-form>"}
        ${"<f-form><template #error-message>Error</template></f-form>"}
        ${'<f-form display-error="maybe"><button type="submit">Submit</button></f-form>'}
        ${'<f-form><template #tooltip>Tooltip</template><button type="submit">Submit</button></f-form>'}
    `("$html should be invalid", ({ html }) => {
        expect.assertions(3);
        let catchedError;

        try {
            expect(html).toHTMLValidate();
        } catch (error) {
            catchedError = error;
        } finally {
            expect(catchedError).toBeDefined();
            expect(catchedError).toMatchSnapshot();
        }
    });
});
