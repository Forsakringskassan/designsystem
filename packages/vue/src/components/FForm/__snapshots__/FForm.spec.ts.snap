// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html-validate <f-form display-error="maybe"><button type="submit">Submit</button></f-form> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  Attribute "display-error" has invalid value "maybe" [attribute-allowed-values]]
`;

exports[`html-validate <f-form></f-form> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <f-form> element must have a submit button [wcag/h32]]
`;

exports[`html-validate <f-form><template #error-message>Error</template></f-form> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <f-form> element must have a submit button [wcag/h32]]
`;

exports[`html-validate <f-form><template #tooltip>Tooltip</template><button type="submit">Submit</button></f-form> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <f-form> component has no slot "tooltip" [vue/available-slots]]
`;

exports[`snapshots should match snapshot when pristine 1`] = `
<form
  class="form"
  id="test-id"
  novalidate=""
>
  <!--v-if-->
  
  <div
    class="text-field"
  >
    <div
      class="i-width-sm-12"
    >
      <f-label-stub
        class=""
        for="element1"
      />
    </div>
    <div
      class="text-field__input-wrapper i-width-sm-12"
    >
      
      
      <div
        class="text-field__icon-wrapper"
      >
        <input
          aria-invalid="false"
          class="text-field__input"
          data-required=""
          data-validation=""
          id="element1"
          required=""
          type="text"
        />
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
      </div>
      
      
    </div>
    <!--v-if-->
  </div>
  <div
    class="text-field"
  >
    <div
      class="i-width-sm-12"
    >
      <f-label-stub
        class=""
        for="element2"
      />
    </div>
    <div
      class="text-field__input-wrapper i-width-sm-12"
    >
      
      
      <div
        class="text-field__icon-wrapper"
      >
        <input
          aria-invalid="false"
          class="text-field__input"
          data-required=""
          data-validation=""
          id="element2"
          required=""
          type="text"
        />
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
      </div>
      
      
    </div>
    <!--v-if-->
  </div>
  
</form>
`;

exports[`snapshots should match snapshot with errors 1`] = `
<form
  class="form"
  id="test-id"
  novalidate=""
>
  <f-message-box-stub
    banner="false"
    layout="standard"
    providescreenreadercontext="true"
    tabindex="-1"
    type="error"
  />
  
  <div
    class="text-field text-field--error"
  >
    <div
      class="i-width-sm-12"
    >
      <f-label-stub
        class=""
        for="element1"
      />
    </div>
    <div
      class="text-field__input-wrapper i-width-sm-12"
    >
      
      
      <div
        class="text-field__icon-wrapper"
      >
        <input
          aria-invalid="true"
          class="text-field__input"
          data-required=""
          data-validation=""
          id="element1"
          required=""
          type="text"
        />
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
      </div>
      
      
    </div>
    <!--v-if-->
  </div>
  <div
    class="text-field text-field--error"
  >
    <div
      class="i-width-sm-12"
    >
      <f-label-stub
        class=""
        for="element2"
      />
    </div>
    <div
      class="text-field__input-wrapper i-width-sm-12"
    >
      
      
      <div
        class="text-field__icon-wrapper"
      >
        <input
          aria-invalid="true"
          class="text-field__input"
          data-required=""
          data-validation=""
          id="element2"
          required=""
          type="text"
        />
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
        <!--v-if-->
      </div>
      
      
    </div>
    <!--v-if-->
  </div>
  
</form>
`;
